src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs

```diff
@@ -0,0 +1,187 @@
+        {
+            Interlocked.CompareExchange(ref _reusableThreadPoolValueTaskSource, source, null);
+        }
+
+        /// <summary>
+        /// A reusable <see cref="IValueTaskSource"/> implementation that
+        /// queues asynchronous <see cref="RandomAccess"/> operations to
+        /// be completed synchronously on the thread pool.
+        /// </summary>
+        internal sealed class ThreadPoolValueTaskSource : IThreadPoolWorkItem, IValueTaskSource<int>, IValueTaskSource<long>
+        {
+            private enum Operation : byte
+            {
+                None,
+                Read,
+                Write,
+                ReadScatter,
+                WriteGather
+            }
+
+            private ManualResetValueTaskSourceCore<long> _source;
+            private readonly SafeFileHandle _fileHandle;
+            private Operation _operation = Operation.None;
+
+            // These fields store the parameters for the operation.
+            // The first two are common for all kinds of operations.
+            private long _fileOffset;
+            private CancellationToken _cancellationToken;
+            // Used by simple reads and writes. Will be unsafely cast to a memory when performing a read.
+            private ReadOnlyMemory<byte> _singleSegment;
+            // Used by vectored reads and writes. Is an IReadOnlyList of either Memory or ReadOnlyMemory of bytes.
+            private object? _multiSegmentCollection;
+
+            internal ThreadPoolValueTaskSource(SafeFileHandle fileHandle)
+            {
+                _fileHandle = fileHandle;
+                _source.RunContinuationsAsynchronously = true;
+            }
+
+            private long GetResultAndRelease(short token)
+            {
+                try
+                {
+                    return _source.GetResult(token);
+                } finally
+                {
+                    _source.Reset();
+                    _fileHandle.TryToReuse(this);
+                }
+            }
+
+            public ValueTaskSourceStatus GetStatus(short token) => _source.GetStatus(token);
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) =>
+                _source.OnCompleted(continuation, state, token, flags);
+            int IValueTaskSource<int>.GetResult(short token) => (int) GetResultAndRelease(token);
+            long IValueTaskSource<long>.GetResult(short token) => GetResultAndRelease(token);
+
+            void IThreadPoolWorkItem.Execute()
+            {
+                Debug.Assert(_operation >= Operation.Read && _operation <= Operation.WriteGather);
+
+                long result = 0;
+                try {
+                    bool notifyWhenUnblocked = false;
+                    try
+                    {
+                        // This is the operation's last chance to be cancelled.
```

Comment:
This highlights one advantage the task-based solution had: if you requested the operation be canceled while the task was still waiting to run, it would cancel immediately and remove the task from the thread pool's queue.  We could employ an additional interlocked that a canceler would signal to determine whether it had ownership to cancel the operation, with the execution performing a similar interlocked check, but then if cancellation won that thread would need to complete the IVTS, and it probably shouldn't do so as part of the calling thread, at which point we'd need to queue the cancellation (assuming we go with RunContinuationsAsynchronously==false), and thus we wouldn't really gain anything... plus it wouldn't address the issue of the work item remaining in the queue and consuming resources unnecessarily.

But, on the whole I think the PR still makes the right tradeoff, as cancellation should be relatively rare.

Reply:
> remove the task from the thread pool's queue
I might be wrong but that was only going to happen if the task was scheduled in a local thread pool queue and cancelled by the same thread. The first was not happening either way because we weren't passing `TaskCreationOptions.PreferFairness` to `Task.Factory.StartNew`.

Reply:
It's the other way around: Tasks are scheduled to the local queue by default, and PreferFairness causes them to go to the global queue.